modules/402-ingress-nginx/hooks/alert_deprecated_geoip_version.go

```diff
@@ -0,0 +1,180 @@
+	ingressAnnotationPrefix = "nginx.ingress.kubernetes.io/"
+
+	eolVersion = semver.MustParse("1.10")
+
+	geoipVarsRegexp = regexp.MustCompile(`\$geoip_(country_(code3|code|name)|area_code|city_continent_code|city_country_(code3|code|name)|dma_code|latitude|longitude|region|region_name|city|postal_code|org)([^_a-zA-Z0-9]|$)+`)
+
+	// objectBatchSize - how many ingress objects to list from k8s at once
+	objectBatchSize = int64(30)
+	// fetchSecretsInterval pause between fetching ingress objects from apiserver
+	fetchSecretsInterval = 1 * time.Second
+)
+
+type ControllerVersion struct {
+	Name    string
+	Version string
+}
+
+func inletHostWithFailoverFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	controller := ControllerVersion{}
+	controller.Name = obj.GetName()
+	controllerVersion, ok, err := unstructured.NestedString(obj.Object, "spec", "controllerVersion")
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get controllerVersion field from ingress controller %s: %w", controller.Name, err)
+	}
+
+	if ok {
+		controller.Version = controllerVersion
+	}
+	return controller, nil
+}
+
+func searchForDeprecatedGeoip(input *go_hook.HookInput, dc dependency.Container) (err error) {
+	kubeClient := dc.MustGetK8sClient()
+	input.MetricsCollector.Expire(metricsGroup)
+	defaultVersion, err := semver.NewVersion(input.Values.Get("ingressNginx.defaultControllerVersion").String())
+	if err != nil {
+		return fmt.Errorf("couldn't parse defaultControllerVersion as semver: %w", err)
+	}
+
+	controllers := input.Snapshots["controller"]
+
+	// check ingressnginxcontrollers' configs
+	for _, c := range controllers {
+		controller := c.(ControllerVersion)
+		var cVer *semver.Version
+
+		if len(controller.Version) == 0 {
+			cVer = defaultVersion
+		} else {
+			cVer, err = semver.NewVersion(controller.Version)
+			if err != nil {
+				return fmt.Errorf("couldn't parse controller's version as semver: %w", err)
+			}
+		}
+
+		if cVer.Compare(eolVersion) >= 0 {
+			continue
+		}
+
+		err := checkControllerConfigMaps(kubeClient, input.MetricsCollector, controller.Name)
+		if err != nil {
+			return fmt.Errorf("couldn't check %s controller's configMaps: %w", controller.Name, err)
+		}
+	}
+
+	// check ingress objects' annotations
+	var next string
+
+	for {
+		ingressList, err := kubeClient.NetworkingV1().Ingresses("").List(context.Background(), metav1.ListOptions{
+			Limit:                objectBatchSize,
+			Continue:             next,
+			ResourceVersion:      "0",
+			ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan,
+		})
+		if err != nil {
+			return fmt.Errorf("couldn't list ingresses: %w", err)
+		}
+		ingressList.GetRemainingItemCount()
+
+		for _, ingress := range ingressList.Items {
+			annotations := ingress.Annotations
+			if len(annotations) == 0 {
+				continue
+			}
+
+			for k, v := range annotations {
+				if strings.HasPrefix(k, ingressAnnotationPrefix) && geoipVarsRegexp.MatchString(v) {
```

Comment:
Why we couldn't use `... && strings.HasPrefix("$geoip_")` ?

Reply:
just to avoid guessing possible variations of different annotations, I suppose. To make sure.